Added BRNN support

SeanNaren opened the pull request:
Hey I've added BLSTM support, hopefully it looks good! I've had to expose some methods in RNN to inherit. Hopefully it's useful, let me know of any feedback.

Commit:
Changed local/global functions to refer to self for inheritance

Commit:
Added BLSTM based on RNN implementation

Commit:
added BLSTM to init

Commit:
removed self call in makeContiguous

borisfom commented:
Looks good. Can you add some tests, too ?

SeanNaren commented:
Good idea but not sure how to test the RNN script, is there a way that we could get the bias and weights of the gate? How would you suggest testing the two modules?

ngimel commented:
cudnn provides cudnnGetRNNLinLayerMatrixParams and cudnnGetRNNLinLayerBiasParams functions for getting pointers to biases and weights.

SeanNaren commented:
I'm trying to replicate the sample as to act as the test however I can't figure out how to create the double pointer required for the cudnnGetRNNLinLayerMatrixParams method for the last arg:

```
float *linLayerMat;

         cudnnErrCheck(cudnnGetRNNLinLayerMatrixParams( cudnnHandle,
                                                        rnnDesc,  
                                                        layer,
                                                        xDesc, 
                                                        wDesc, 
                                                        w,
                                                        linLayerID,  
                                                        linLayerMatDesc, 
                                                        (void**)&linLayerMat));
```

How would I create the double pointer? My attempt may also help:

```
local ptr = ffi.new("float*")

local outputDesc = rnn:createFilterDescriptors(1)
errcheck('cudnnGetRNNLinLayerMatrixParams',
    cudnn.getHandle(),
    rnn.rnnDesc[0],
    0,
    rnn.xDescs,
    rnn.wDesc[0],
    rnn.weight:data(),
    0,
    outputDesc[0],
    ptr) -- How to cast float* to void**?
```

ngimel commented:
This stackoverflow question might help
http://stackoverflow.com/questions/14011598/how-to-pass-a-pointer-to-luajit-ffi-to-be-used-as-out-argument

SeanNaren commented:
Thanks! Managed to cast it using ffi to get the pointer like such, but not sure how to access the actual data from the pointer. In the C RNN example all values in the matrix/bias are initialised to a set value, would I be able to somehow replicate that in lua?

```
 local linLayerMatDesc = rnn:createFilterDescriptors(1)
        local matrixPointer = ffi.new("float*[1]")
        errcheck('cudnnGetRNNLinLayerMatrixParams',
            cudnn.getHandle(),
            rnn.rnnDesc[0],
            layer,
            rnn.xDescs,
            rnn.wDesc[0],
            rnn.weight:data(),
            layerId,
            linLayerMatDesc[0],
            ffi.cast("void**", matrixPointer)) -- How to access data stored at this pointer?
```

ngimel commented:
You'll have to get weight size from `linLayerMatDesc[0]`, calculate offset as
`offset = matrixPointer[0] - rnn.weight:data()`, and then do something like

```
wTensor = torch.Tensor(rnn.weight:storage(), offset,size)
wTensor:fill(1)
```

SeanNaren commented:
Awesome, and for the bias values would the same approach work? The command is nearly identical:

```
 local linLayerBiasDesc = rnn:createFilterDescriptors(1)
        local biasPointer = ffi.new("float*[1]")
        errcheck('cudnnGetRNNLinLayerBiasParams',
            cudnn.getHandle(),
            rnn.rnnDesc[0],
            layer,
            rnn.xDescs,
            rnn.wDesc[0],
            rnn.weight:data(),
            layerId,
            linLayerBiasDesc[0],
            ffi.cast("void**", biasPointer))
```

ngimel commented:
Yes, this should work for biases too.

Commit:
Added rnn test (for basic RELU)

Commit:
Removed c sum check

Commit:
Removed print statements

SeanNaren commented:
Added a test for RNN that uses the checksums taken from the cudnn RNN C sample. The checksums for a few of the backward passes are not correct, any help would be great! Will try to fix them as well.

Commit:
Fixed call to input grads

Commit:
Added test for all RNN types

Commit:
Added description

SeanNaren commented:
The test now works for RELU/TANH/LSTM/GRU. One failing test (the weight checksum for ReLU is slightly off, not entirely sure why).

ngimel commented:
Great, thanks!

SeanNaren commented:
Just as a side note would it make more sense to add the LSTM/GRU RNNs as separate modules like a BLSTM?

Commit:
Changed tolerance to fix weight difference, small comment change

Commit:
Exposed rnn variable to make easier to replace

ngimel commented on test/test_rnn.lua:
```diff
@@ -0,0 +1,178 @@
+            local dataType = 'CUDNN_DATA_FLOAT'
+            local format = 'CUDNN_TENSOR_NCHW'
+            local nbDims = torch.IntTensor(1)
+            local filterDimA = torch.IntTensor({ 1, 1, 1 })
```
You can remove hard-coded 3 dimensions by 

```
local minDim=3
local filterDimA=torch.ones(minDim)
```

replace '3' argument to GetFilterNdDescriptor call with `minDim`, and replace `filterDimA[1] * filterDimA[2] * filterDimA[3]` with `filterDimA:prod()`

Commit:
Added base test for BLSTM

Commit:
Revert "Added description"

This reverts commit 11fe8331da8040f1f40a1a1b7966e06057a10c5b.

Commit:
Fixed change of RNN module name

ngimel commented on BLSTM.lua:
```diff
@@ -0,0 +1,67 @@
+local BLSTM, parent = torch.class('cudnn.BLSTM', 'cudnn.RNN')
+local errcheck = cudnn.errcheck
+
+function BLSTM:__init(inputSize, hiddenSize, numLayers)
```
Does it make sense to add self.numDirections to base RNN class (set it to 1 for uni-directional, and 2 for bi-directional), and use this var to avoid duplicating all of __init, resizeOutput, resizeHidden and resetIODescriptors?
That way, for BLSM you will just have to call parents __init, and set self.numDirections, self.bidirectional and self.mode in __init. resizeOutput, resizeHidden and resetIODescriptors can be reused for bidirectional nets w/o changes.

ngimel commented:
It makes sense to have separate modules for GRU/LSTM, but see my previous comment - we should try to reduce code duplication.

SeanNaren commented:
I'm running into issues with the test and I feel like it is due to the assertions made (not knowing it is bi-directional). I think your suggestion may fix this, will revert a few things and get it going!

Commit:
Added comment removed BLSTM test

Commit:
Added numDirections param to RNN

Commit:
Added numDirections param instead of duplicated methods

Commit:
Removed *2 on gradInput

Commit:
Removed hardcoded 3 dimension from cudnn call

Commit:
Added BLSTM test

SeanNaren commented:
Great thanks! That should be the conversion to RNN handling a bit more, looks much cleaner imo. Having an issue with implementing the bidirectional scale in the test, not sure if the scaling is being applied correctly the vars:

```
    -- Set hx/cx/dhy/dcy data to 1s.
    rnn.hiddenInput = torch.CudaTensor(numberOfLayers, miniBatch, hiddenSize * biDirectionalScale):fill(1)
    rnn.cellInput = torch.CudaTensor(numberOfLayers, miniBatch, hiddenSize * biDirectionalScale):fill(1)
    rnn.gradHiddenOutput = torch.CudaTensor(numberOfLayers * biDirectionalScale, miniBatch, hiddenSize):fill(1)
    rnn.gradCellOutput = torch.CudaTensor(numberOfLayers * biDirectionalScale, miniBatch, hiddenSize):fill(1)

    local testOutputi = rnn:forward(input)
    -- gradInput set to 1s.
    local gradInput = torch.CudaTensor(seqLength, miniBatch, hiddenSize * biDirectionalScale):fill(1) 

```

SeanNaren commented:
@borisfom Just saw that the R5 branch has been merged, how would you suggest adding RNN/tests etc?

borisfom commented:
@SeanNaren : only non-RNN part was merged. So I propose I will rebase my R5 to upstream R5 today or tomorrow, then we'll merge your PR as soon as we sort the test stuff out (or the other way around).

SeanNaren commented:
@borisfom Sounds good, will try to fix the BRNN test as soon as I can. The C example RNN doesn't seem to give me good ref values for BLSTM (Nans for some reason) but using a RELU seems to be fine, so will try create a test on that!

EDIT: @ngimel having difficulties understanding what the shape should be for hy,cy when bidirectional is applied, should they both be: 

```
self.numLayers, self.miniBatch, self.hiddenSize * 2 
```

or 

```
self.numLayers * 2, self.miniBatch, self.hiddenSize
```

and should hx,cx be of shape for bidirectional:

```
numberOfLayers * 2, miniBatch, hiddenSize
```

Thanks for your help!

ngimel commented:
self.numLayers \* 2, self.miniBatch, self.hiddenSize 
is correct for hx/hy etc.
The “layer” terminology gets a bit confusing when you’re bidirectional. Is it two layers being concatenated, or one bidirectional layer.

SeanNaren commented:
@ngimel Agreed I am quite baffled as to what is happening when direction is set to bidirectional now. Another issue is how the RNN C example doesn't give me reliable results if I set bidirectional to true, it gives me different checksums each time. Any idea as to why this happens and how we could test bidirectional?

EDIT: it seems after a few NANs and strange results the checksums do infact stay the same.

Commit:
Fixed resize of hidden tensors

Commit:
Added numDirections to assertions

Commit:
Added assertion check to accGradParams

Commit:
Added BRNN test

SeanNaren commented:
Still having trouble with the final BRNN test, but once this is complete it should be good to go. If anyone could help me try find out why the test is failing it would be great!

Commit:
Reverted change of class name

ngimel commented:
It looks like there bugs in cudnn for bidirectional (in cudnnGetRNNLinLayerMatrixParams), which means currently you can not initialize weights properly, C sample fails initcheck currently. I'll let you know once I figure a workaround to test your integration, bugs should be fixed in final version.

SeanNaren commented:
Awesome thanks!

elbamos commented:
@ngimel Is that bug confined to testing or is the test accurately failing?  (In other words, should I wait to try this until the bidirectional bug is fixed.)

ngimel commented:
With high probability bug is confined to cudnnGetRNNLinLayerMatrixParams/cudnnGetRNNLinLayerBiasParams, so as long as you are not using those functions to access/fill parameters and biases you should be fine. It pretty much limits you to same optimizer parameters for all layers and initially filling weights and biases from the same distribution.

elbamos commented:
I'm curious why RNN and BLSTM ask for tensors where the first dimension is sequence rather than batch.  Is this a cudnn thing and we just have to get use to transposing before and after?

elbamos commented:
FYI - Performance of an MD-RNN using this, compared to one made using another common torch RNN implementation, is about 20x better.  :)

SeanNaren commented:
I think it is to follow the norm of the nn library, which for the most part asks for a batch x time x inputdim input. This is also the same as torch-rnn (N x T x D) and seems to only be different in Element Research rnn. 

EDIT: whoops I am very wrong, cudnn does in fact take the input as (T x N xD) as said below, just a silly mistake by me, sorry!

ngimel commented:
@SeanNaren attached patch should give you deterministic results for bidirectional, please let me know if it does not. Note that for bidirectional we are now filling the whole weight matrix with the same value, rather than filling weights with values depending on the weight matrix size, and biases with the different value. This will change back once the bug in cudnnGetRNNLinLayerMatrixParams is fixed. 
[patch.txt](https://github.com/borisfom/cudnn.torch/files/221281/patch.txt)
@elbamos, @SeanNaren  - cudnn RNN expects inputs in the form of time (seqLength) x miniBatch x inpudim, rather than minibatch x time x inputdim that many other packages do to be able to support variable sequence length within minibatch, and to fuse inputs which helps smaller minibatches (see https://devblogs.nvidia.com/parallelforall/optimizing-recurrent-neural-networks-cudnn-5/, optimization 5)

Commit:
Added ReLU/tanh/LSTM/GRU bidirectional tests

SeanNaren commented:
@ngimel Thanks dude worked great! I think that fixes the tests, added a test for each type of layer in bidirectional format. They all pass now but one thing to mention is they take quite a bit of time to run. Other than that the package should be good to go!

What do you guys think about me adding similar modules for GRU/LSTM, could throw those in. How should I handle BRNN formats for these?

Nice job guys! Once everything is done I guess a rebase would be good.

ngimel commented:
Tests taking a long time to run must be because of allocations. Once allocations are out of the way, subsequent calls to :forward() and :backward() should be fast, and should generally match times printed out by the RNN_example. 
Adding GRU/LSTM modules would be great, thanks for the offer! I did not quite get your question about BRNN formats handling?

Commit:
Updated modules in tests

Commit:
Added GRU and LSTM

Commit:
Added GRU/LSTM to init

Commit:
Removed direction set

Commit:
Changed init

Commit:
Fixed module names

SeanNaren commented:
Great stuff I've added the GRU/LSTM modules and edited the tests to use these modules instead. 

On the topic of BRNN, we only have a module for BLSTM. Should we support BGRU/BRNN as well? How would you suggest supporting those modules?

elbamos commented:
@SeanNaren Great work!  Three suggestions:
1.  Provide some guidance, either in documentation or programmatically, for isolating for initialization, access, etc., each part of the weight matrices.  
2.  Add a parameter for whether the input is in tbf or btf, and automatically transpose btf to tbf.  The other layers in the network are almost certain to have minibatch as the first index, so everyone's going to be bracketing this module with boilerplate transpositions.  For that reason, I suggest that btf should be the assumed format if the parameter is left off, kind of like the way nn.Reshape works.
3.  Does cu-dnn not let us have any control over activation functions?

ngimel commented:
For simple RNNs cudnn provides ReLU and TANH activations, LSTMs are implemented according to the paper. If you want something else, now is the time to speak up.

SeanNaren commented:
@elbamos thanks for your feedback, when it comes to initialising the weights/bias I had to do it via descriptors etc to change the individual values. I'm not sure how important this is at the current stage but if @ngimel thinks its important I could start implementing these.

On the topic of the parameter, sounds like a good idea will implement it the same way as described [here](https://github.com/Element-Research/rnn/pull/207) and see what you guys think.

Commit:
added batchfirst param

Commit:
Put tranpose at top of method call

Commit:
Added batchFirst to accGrad

Commit:
Fixed transpose name

Commit:
Fixed transpose name for grads

Commit:
Added all transpose operations

Commit:
Added batchFirst to test params

SeanNaren commented:
Added batchFirst to the RNN, let me know what you guys think! 

EDIT: should we support RNN ReLU/Tanh as separate modules? Only thing I'm concerned is about is that after setting self.mode to either,  you have to reset the rnndescriptor (a call to reset()). Does this warrant a separate module?

Commit:
Added small comment to clarify batchFirst

Commit:
Added description of recurrent modules

ngimel commented:
@SeanNaren, Tanh, ReLu and Sigmoid activations all have their own separate modules, so I don't see why Tanh or ReLU RNNs shouldn't. Great work and thanks for your help! 
@elbamos, @SeanNaren, do you guys think we should handle saving hidden outputs/using them as hidden inputs somewhat more consistently than it is done now in the base RNN class? Say, have functions for copying hy to hx, and forgetting hx? Have an option to not output hy (cudnn allows this)? That would provide some (small) memory savings if you don't have to unroll beyond a single call to cudnn.

elbamos commented:
@ngimel Maybe other people would, but I do not need the ability to drill-down into hx and hy.   I would, though, benefit if we had a choice between keeping all hidden outputs versus only the last output from the sequence, but not in a B-RNN.  Is that what you meant?

elbamos commented:
I'm seeing this with the current version when I try to create a new BLSTM layer.  The BLSTM layer I created last night continues to function and train.

```
th> cnn:add(cudnn.BLSTM(7040, 7040 * 16, 2, true))
cuda runtime error (77) : an illegal memory access was encountered at /tmp/luarocks_cutorch-scm-1-3384/cutorch/lib/THC/generic/THCStorage.c:147
stack traceback:
        [C]: at 0x7f5cfb1f3210
        [C]: at 0x7f5cfb1f8170
        [C]: in function 'CudaTensor'
        /usr/local/share/lua/5.1/cudnn/RNN.lua:114: in function 'resetDropoutDescriptor'
        /usr/local/share/lua/5.1/cudnn/RNN.lua:47: in function 'reset'
        /usr/local/share/lua/5.1/cudnn/RNN.lua:33: in function '__init'
        /usr/local/share/lua/5.1/cudnn/BLSTM.lua:4: in function '__init'
        /usr/local/share/lua/5.1/torch/init.lua:91: in function </usr/local/share/lua/5.1/torch/init.lua:87>
        [C]: in function 'BLSTM'
        [string "_RESULT={cnn:add(cudnn.BLSTM(7040, 7040 * 16,..."]:1: in main chunk
        [C]: in function 'xpcall'
        /usr/local/share/lua/5.1/trepl/init.lua:651: in function 'repl'
        /usr/local/lib/luarocks/rocks/trepl/scm-1/bin/th:199: in main chunk
        [C]: at 0x00405ea0
```

Commit:
Added separate RNN modules

Commit:
Added modules to init

SeanNaren commented:
@elbamos that's a big hidden dim you have there, I don't think I can fit that onto my memory in the first place to test! It works alright with a smaller inputdim on my end however. 

Added separate modules for Tanh and for ReLU. @ngimel Sounds great but I'm sure you are much more informed than me, so any guidance on how to approach this would be great :) 

EDIT: should've done this some time ago, added you as as collab ngimel feel free to change anything you see fit!

EDIT2: Don't want to spam too much but tests are now much faster. Should've realised earlier it was the manual for loops that were taking up all our time...

Commit:
Added RNNReLU/Tanh description

Commit:
Added batchFirst to params

Commit:
Changed module calls in tests

Commit:
Use torch sum instead of manual loop

SeanNaren commented:
@borisfom I think the support is mostly done and tests are looking good, how would you suggest us merging the stuff here and RNN into main? Shall I rebase onto R5 into a new branch and open a new request with that branch into R5?

borisfom commented:
Please hold on - I am rebasing my R5 to upstream R5 (nontrivial exercise due to selective merge they did earlier). This will likely result in a new branch - then I will ask you to do a PR against that branch.

ngimel commented:
@SeanNaren Thanks! Looks like neither you nor @elbamos need more flexibility for hidden layers, so we can leave that till later date. You guys serve as proxy for real RNN users ;-) 
@elbamos - currently you can not keep all your hidden outputs if you are processing your whole sequence in a single call to cudnn. If you really need it, you need to construct your network as a series of calls to cudnn, one for every time step - pretty much like its done in other torch rnn packages. Then you'll have all your hidden outputs (and you'll also lose performance optimizations beyond "Optimizing a single iteration" listed in the blogpost. Though "Pre-Transposing the Weight Matrix" possibly can be used even for this case, but it will require changing APIs, so not in this cudnn version:-). Let me know if you think this is an important case, may be we should have an API call for changing weight matrix layout).

borisfom commented:
@SeanNaren : actually, please nevermind, I will merge the PR before rebase.
I will have to do some squashing (rebase -i HEAD~n) before merging it back to soumith's branch (ideally down to 2 or 3 commits) : please let me know if you would rather do it yourself on your branch now.

SeanNaren commented:
@ngimel, I think I see what you mean now, if it is a feature people want once the PR is merged then I'll be more than happy to take a look :)
@borisfom I'll help you out on the part, give me some time and I'll rebase to a new branch ASAP.

borisfom commented:
@SeanNaren: yes the best course would be for you to create a new branch from current one, and then squash that one to very few commits, then send a new PR from that branch (unless you can change the branch in the PR directly)

Commit:
Put resizing of tensors in a better place

SeanNaren commented:
@borisfom, Right [here](https://github.com/SeanNaren/cudnn.torch/tree/R5Rebase) is the rebased branch, hopefully its alright just being 1 commit, let me know what you think! 

If its good let me know what branch you want the PR to.

borisfom commented:
SeanNaren: yes one commit is perfect! Please do PR against branch R5-re.

SeanNaren commented:
@borisfom @ngimel thought this may be a suitable place to ask, how difficult would it be to support summation of the output layers rather than concat for BRNNs?

ngimel commented:
You can always add your outputs outside of cuDNN. We went with concatenation rather than addition because it preserves the information and leaves the user free to postprocess the output the way user sees fit. There is no overlap between layers in a bidirectional network, so if you split layers and add the outputs yourself, you should not loose much performance.

SeanNaren commented:
Awesome thanks!
